File: algorithms/2_medium/C/00787_cheapest_flights_within_k_stops/bellman_ford.c
#define MAX_ALLOWED_PRICE 10000
#define MAX_ALLOWED_N 100
#define MAX_ALLOWED_FLIGHTS ((MAX_ALLOWED_N * (MAX_ALLOWED_N - 1)) / 2)
#define INF (INT_MAX / 2)

static inline int min_int(const int a, const int b)
{
    return a < b ? a : b;
}

int findCheapestPrice(int n, int** flights, int flights_len, int* flights_col_lem, int src, int dst, int k)
{
    assert(k >= 0 && k < n && n <= MAX_ALLOWED_N && flights_len <= MAX_ALLOWED_FLIGHTS);
    assert(src >= 0 && src < n && dst >= 0 && dst < n && src != dst);

    int* buff = malloc((size_t)(2 * n) * sizeof(*buff));
    assert(buff != NULL);
    int* curr = &buff[0];
    int* next = &buff[n];
    for (int i = 0; i < n; i += 1) {
        curr[i] = INF;
        next[i] = INF;
    }
    curr[src] = 0;
    next[src] = 0;

    for (int stops = 0; stops <= k; stops += 1) {
        for (int i = 0; i < flights_len; i += 1) {
            const int* f = flights[i];
            const int u = f[0];
            const int v = f[1];
            const int w = f[2];
            assert(u >= 0 && u < n && v >=0 && v < n && w >= 1 && w <= MAX_ALLOWED_PRICE);
            next[v] = min_int(next[v], curr[u] + w);
        }
        int* temp = curr;
        curr = next;
        next = temp;
    }

    const int res = curr[dst] == INF ? -1 : curr[dst];
/* cleanup: */
    free(buff);
    return res;
}

File: algorithms/2_medium/C/00787_cheapest_flights_within_k_stops/bellman_ford.rs
impl Solution {
    const MAX_ALLOWED_PRICE: i32 = 10_000;
    const MAX_ALLOWED_N: i32 = 100;
    const MAX_ALLOWED_FLIGHTS: i32 = (Self::MAX_ALLOWED_N * (Self::MAX_ALLOWED_N - 1)) / 2;
    const INF: i32 = i32::MAX / 2; 

    pub fn find_cheapest_price(n: i32, flights: Vec<Vec<i32>>, src: i32, dst: i32, k: i32) -> i32 {
        assert!(k >= 0 && k < n && n <= Self::MAX_ALLOWED_N && flights.len() <= Self::MAX_ALLOWED_FLIGHTS as usize);
        assert!(src >= 0 && src < n && dst >= 0 && dst < n && src != dst);

        let mut buff = vec![Self::INF; 2 * (n as usize)];
        let (mut curr, mut next) = buff.split_at_mut(n as usize);
        curr[src as usize] = 0;
        next[src as usize] = 0;

        for _ in 0..=k {
            for f in &flights {
                let (u, v, w) = (f[0], f[1], f[2]);
                assert!(u >= 0 && u < n && v >=0 && v < n && w >= 1 && w <= Self::MAX_ALLOWED_PRICE);
                next[v as usize] = next[v as usize].min(curr[u as usize] + w);
            }
            (curr, next) = (next, curr);
        }

        if curr[dst as usize] == Self::INF { -1 } else { curr[dst as usize] }
    }
}
